io.c

```diff
@@ -3203,67 +3203,87 @@ io_getpartial(int argc, VALUE *argv, VALUE io, int no_exception, int nonblock)
 
 /*
  *  call-seq:
- *     ios.readpartial(maxlen)              -> string
- *     ios.readpartial(maxlen, outbuf)      -> outbuf
+ *    readpartial(maxlen)             -> new_string
+ *    readpartial(maxlen, out_string) -> out_string
  *
- *  Reads at most <i>maxlen</i> bytes from the I/O stream.
- *  It blocks only if <em>ios</em> has no data immediately available.
- *  It doesn't block if some data available.
+ *  Reads up to +maxlen+ bytes from the stream.
  *
- *  If the optional _outbuf_ argument is present,
- *  it must reference a String, which will receive the data.
- *  The _outbuf_ will contain only the received data after the method call
- *  even if it is not empty at the beginning.
+ *  With the single non-negative integer argument +maxlen+ given,
+ *  returns a new string with +maxlen+ bytes from the stream, if available,
+ *  or all available bytes otherwise:
  *
- *  It raises EOFError on end of file.
- *
- *  readpartial is designed for streams such as pipe, socket, tty, etc.
- *  It blocks only when no data immediately available.
- *  This means that it blocks only when following all conditions hold.
- *  * the byte buffer in the IO object is empty.
- *  * the content of the stream is empty.
- *  * the stream is not reached to EOF.
- *
- *  When readpartial blocks, it waits data or EOF on the stream.
- *  If some data is reached, readpartial returns with the data.
- *  If EOF is reached, readpartial raises EOFError.
- *
- *  When readpartial doesn't blocks, it returns or raises immediately.
- *  If the byte buffer is not empty, it returns the data in the buffer.
- *  Otherwise if the stream has some content,
- *  it returns the data in the stream.
- *  Otherwise if the stream is reached to EOF, it raises EOFError.
- *
- *     r, w = IO.pipe           #               buffer          pipe content
- *     w << "abc"               #               ""              "abc".
- *     r.readpartial(4096)      #=> "abc"       ""              ""
- *     r.readpartial(4096)      # blocks because buffer and pipe is empty.
- *
- *     r, w = IO.pipe           #               buffer          pipe content
- *     w << "abc"               #               ""              "abc"
- *     w.close                  #               ""              "abc" EOF
- *     r.readpartial(4096)      #=> "abc"       ""              EOF
- *     r.readpartial(4096)      # raises EOFError
+ *    f = File.new('t.txt')
+ *    f.readpartial(30) # => "This is line one.\nThis is the"
+ *    f.readpartial(30) # => " second line.\nThis is the thi"
+ *    f.readpartial(30) # => "rd line.\n"
+ *    f.eof             # => true
+ *    f.readpartial(30) # Raises EOFError.
+ *
+ *  With integer argument +maxlen+ and string argument +out_string+ given,
+ *  returns +out_string+ with +maxlen+ bytes from the stream, if available,
+ *  or all available bytes otherwise;
+ *  if any bytes are read, they replace the content of +out_string+:
```

Comment:
`out_string` is replaced by the empty string if no bytes are read, unless `maxlen` is `0` (which is inconsistent with `read`, not sure if that is expected or not):

```ruby
s = nil; File.open('ruby.c'){|f| f.read; p f.readpartial(0, s = "x")}; s # => 'x'
s = nil; File.open('ruby.c'){|f| f.read; p f.readpartial(1, s = "x") rescue nil}; s # => ''
```

Reply:
I think it is a bug.

Reply:
This seems obviously wrong.
```ruby
File.open(IO::NULL) {|f| p f.readpartial(0, s = "x"); p s} #=> "x"
```

Reply:
Noted.